core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/CompositeRuinMoveSelector.java

```diff
@@ -0,0 +1,266 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic;
+
+import static org.optaplanner.core.impl.heuristic.move.CompositeMove.buildMove;
+
+import java.util.*;
+
+import org.optaplanner.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ShadowVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.VariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.inverserelation.InverseRelationShadowVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableDemand;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
+import org.optaplanner.core.impl.domain.variable.supply.SupplyManager;
+import org.optaplanner.core.impl.heuristic.move.Move;
+import org.optaplanner.core.impl.heuristic.selector.common.iterator.UpcomingSelectionIterator;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.move.generic.chained.SubChainRuinMove;
+import org.optaplanner.core.impl.heuristic.selector.move.generic.list.ListUnassignMove;
+import org.optaplanner.core.impl.heuristic.selector.value.chained.SubChain;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+
+public class CompositeRuinMoveSelector<Solution_> extends GenericMoveSelector<Solution_> {
+
+    private final EntitySelector<Solution_> leftEntitySelector;
+    private final EntitySelector<Solution_> rightEntitySelector;
+    private final List<GenuineVariableDescriptor<Solution_>> variableDescriptorList;
+    private final Collection<ShadowVariableDescriptor<Solution_>> shadowVariableDescriptors;
+    private final Integer percentageToBeRuined;
+
+    private final HashMap<GenuineVariableDescriptor<Solution_>, SingletonInverseVariableSupply> variableDescriptorToInverseVariableSupply =
+            new HashMap<>();
+    private final boolean chained;
+
+    public CompositeRuinMoveSelector(EntitySelector<Solution_> leftEntitySelector,
+            EntitySelector<Solution_> rightEntitySelector,
+            List<GenuineVariableDescriptor<Solution_>> variableDescriptorList,
+            Collection<ShadowVariableDescriptor<Solution_>> shadowVariableDescriptors, Integer percentageToBeRuined) {
+        assert 0 < percentageToBeRuined && percentageToBeRuined <= 100
+                : "CompositeRuinMoveSelector: percentage to ruin the solution must be greater than 0 and smaller or equal to 100%";
```

Comment:
We generally don't use these JDK assertions. They need to be specifically enabled to work, and nobody ever enables them in their JDK.

Please make this an `IllegalArgumentException` instead.